nestkernel/stimulating_device.h

```diff
@@ -88,27 +90,69 @@ namespace nest
  *
  * @ingroup Devices
  */
-template < typename EmittedEvent >
-class StimulatingDevice : public Device
+class StimulatingDevice : public DeviceNode, public Device
 {
 public:
   StimulatingDevice();
-  StimulatingDevice( StimulatingDevice< EmittedEvent > const& );
-  virtual ~StimulatingDevice()
-  {
-  }
+  StimulatingDevice( StimulatingDevice const& );
+  ~StimulatingDevice() override = default;
 
   /**
    * Determine whether device is active.
    * The argument is the value of the simulation time.
    * @see class comment for details.
    */
-  bool is_active( const Time& ) const;
-  void get_status( DictionaryDatum& d ) const;
+  bool is_active( const Time& ) const override;
+  void get_status( DictionaryDatum& d ) const override;
+  void set_status( const DictionaryDatum& ) override;
+
+  bool has_proxies() const override;
+  Name get_element_type() const override;
+
+  using Device::init_state;
+  using Device::calibrate;
+  using Device::init_buffers;
+  using Node::calibrate;
+
+  void calibrate() override;
 
   //! Throws IllegalConnection if synapse id differs from initial synapse id
   void enforce_single_syn_type( synindex );
 
+  /**
+   * Device type.
+   */
+  enum Type
+  {
+    CURRENT_GENERATOR,
+    SPIKE_GENERATOR,
+    DOUBLE_DATA_GENERATOR,
+    DELAYED_RATE_CONNECTION_GENERATOR,
+  };
+
+  virtual Type
+  get_type() const
+  {
+    throw KernelException( "WORNG TYPE" );
+  };
+  const std::string& get_label() const;
+  virtual void set_data_from_stimulating_backend( std::vector< double > ){};
+  void update( Time const&, const long, const long ) override{};
+
+protected:
+  void set_initialized_() final;
+
+  struct Parameters_
+  {
+    std::string label_;    //!< A user-defined label for symbolic device names.
+    Name stimulus_source_; //!< Origin of the stimulating signal.
+
+    Parameters_();
+    Parameters_( const Parameters_& );
```

Comment:
The copy constructor should be defaulted here, not in the cpp file.

Reply:
Done